src/lro/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+pub trait Operation {
+    type ResponseType;
+    type MetadataType;
+
+    fn name(&self) -> String;
+    fn done(&self) -> bool;
+    fn metadata(&self) -> Option<&wkt::Any>;
+    fn response(&self) -> Option<&wkt::Any>;
+    fn error(&self) -> Option<&rpc::model::Status>;
+}
+
+/// Implements [Operation] using [longrunning::model::Operation].
+pub struct TypedOperation<R, M> {
+    inner: longrunning::model::Operation,
+    response: std::marker::PhantomData<R>,
+    metadata: std::marker::PhantomData<M>,
+}
+
+impl<R, M> TypedOperation<R, M> {
+    pub fn new(inner: longrunning::model::Operation) -> Self {
+        Self {
+            inner,
+            response: PhantomData,
+            metadata: PhantomData,
+        }
+    }
+}
+
+impl<R, M> Operation for TypedOperation<R, M> {
+    type ResponseType = R;
+    type MetadataType = M;
+
+    fn name(&self) -> String {
+        self.inner.name.clone()
+    }
+    fn done(&self) -> bool {
+        self.inner.done
+    }
+    fn metadata(&self) -> Option<&wkt::Any> {
+        self.inner.metadata.as_ref()
+    }
+    fn response(&self) -> Option<&wkt::Any> {
+        use longrunning::model::operation::Result;
+        self.inner.result.as_ref().and_then(|r| match r {
+            Result::Error(_) => None,
+            Result::Response(r) => Some(r),
+            _ => None,
+        })
+    }
+    fn error(&self) -> Option<&rpc::model::Status> {
+        use longrunning::model::operation::Result;
+        self.inner.result.as_ref().and_then(|r| match r {
+            Result::Error(rpc) => Some(rpc),
+            Result::Response(_) => None,
+            _ => None,
+        })
+    }
+}
+
+/// The trait implemented by LRO helpers.
+pub trait Poller<R, M> {
+    fn poll(&mut self) -> impl Future<Output = Option<PollingResult<R, M>>>;
+}
+
+/// An implementation of `Poller` based on closures.
+///
+/// Thanks to this implementation, the code generator (`sidekick`) needs to
+/// produce two closures: one to start the operation, and one to query progress.
+///
+/// Applications should not need to create this type, or use it directly. It is
+/// only public so the generated code can use it.
+///
+/// # Parameters
+/// - `O` - the operation type. Typically this is a `TypedOperation<R, M>`, so
+///   it encodes the types of the response and metadata.
+/// - `S` - the start closure. Starts a LRO. This implementation expects that
+///   all necessary parameters, and request options, including retry options
+///   are captured by this function.
+/// - `SF` - the type of future returned by `S`.
+/// - `Q` - the query closure. Queries the status of the LRO created by `start`.
+///   It receives the name of the operation as its only input parameter. It
+///   should have captured any stubs and request options.
+/// - `QF` - the type of future returned by `Q`.
+pub struct PollerImpl<O, S, SF, Q, QF>
+where
+    O: Operation,
+    S: FnOnce() -> SF + Send + Sync,
```

Comment:
I think the reason for making this take a closure is mainly so we can honor any request options -- is that correct? If so, should we have an API similar to our clients where they can just be explicitly passed in? That way you would not need to capture any state.

What I am trying to tease out is: Is this more of a helper utility or is this a lightweight veneer over LROs that anyone could pickup and use. If the former I think this is fine, but if the later I would say this would be harder for someone to pickup and use.

Reply:
Thanks for the explanation, that makes sense to me.